src/rustc-driver.md

```diff
@@ -0,0 +1,66 @@
+# The Rustc Driver
+
+The [`rustc_driver`] is essentially `rustc`'s `main()` function. It acts as
+the glue for running the various phases of the compiler in the correct order,
+managing state such as the [`CodeMap`] \(maps AST nodes to source code),
+[`Session`] \(general build context and error messaging) and the [`TyCtxt`]
+\(the "typing context", allowing you to query the type system and other cool
+stuff). The `rustc_driver` crate also provides external users with a method
+for running code at particular times during the compilation process, allowing
+third parties to effectively use `rustc`'s internals as a library for
+analysing a crate.
+
+For those using `rustc` as a library, the `run_compiler()` function is the main
+entrypoint to the compiler. Its main parameters are a list of command-line
+arguments and a reference to something which implements the `CompilerCalls`
+trait. A `CompilerCalls` creates the overall `CompileController`, letting it
+govern which compiler passes are run and attach callbacks to be fired at the end
+of each phase.
+
+From `rustc_driver`'s perspective, the main phases of the compiler are:
+
+1. *Parse Input:* Initial crate parsing
+2. *Configure and Expand:* Resolve `#[cfg]` attributes and expand macros
```

Comment:
Name resolution is part of this phase

Reply:
Ooh, I didn't know that! I though name resolution happened some time after the `after_hir_lowering` and before `after_analysis` gets called. 

My [dyna-bindgen] tool was doing analysis at the `after_hir_lowering` stage and I couldn't figure out how to get the fully qualified name of a function arguments type. I thought I needed to wait until later and use the `TyCtxt` to make queries.

[dyna-bindgen]: https://gitlab.com/Michael-F-Bryan/dyna-bindgen/